internal/consensus/msgs.go

```diff
@@ -220,9 +220,11 @@ type VoteMessage struct {
 
 func (*VoteMessage) TypeTag() string { return "tendermint/Vote" }
 
-// ValidateBasic performs basic validation.
+// ValidateBasic checks whether the vote within the message is well-formed.
 func (m *VoteMessage) ValidateBasic() error {
-	return m.Vote.ValidateBasic()
+	// Here we validate votes with vote extensions, since we require vote
+	// extensions to be sent in vote messages during consensus.
+	return m.Vote.ValidateWithExtension()
```

Comment:
I think the answer is "yes", but just to confirm: Do we expect a signature on a prevote, even though we ourselves would not sign a prevote?

Reply:
No, not that we've explicitly said it anywhere, but in my understanding vote extensions are exclusive to precommits. I'm not totally sure what the behavior should be if we see extensions in prevotes - my gut sense is that we should reject the votes, but we could just as easily ignore them.

Reply:
Also, the `ValidateWithExtension` method automatically checks now whether the vote's a prevote or a precommit, and validates accordingly.

Reply:
Ah that makes sense. But then I wonder: if we always expect a valid extension on a precommit, and never do on a prevote, maybe we should just have one validate method that either does (for precommits) or doesn't (for prevotes) check the extension?

Reply:
Right now the check happens within the `ValidateWithExtension` method I'm suggesting to add to `Vote`, which internally ensures that prevotes _don't_ have extensions and that precommits _do_. So I think it's still important to do the vote extension check for prevotes (specifically to reject prevotes that include extensions, as that should be considered a malformed vote).

Reply:
> Right now the check happens within the `ValidateWithExtension` method I'm suggesting to add to `Vote`, which internally ensures that prevotes _don't_ have extensions and that precommits _do_. So I think it's still important to do the vote extension check for prevotes (specifically to reject prevotes that include extensions, as that should be considered a malformed vote).

Makes sense. Given that a `Vote` has a `Type` field to say whether it's a prevote or a precommit, could we not more simply have one unified "validate" that doesn't require the caller to know whether or not the extension is being checked? 

I guess I'm trying to understand when the caller would ever want to "validate this vote ignoring whether its extensions are sensible" (except maybe a test?)

Reply:
Oh, so precommits won't _always_ have vote extensions. When they're reconstructed from commit signatures we won't have vote extensions.

There are therefore two "areas" of code to which we need to cater here: a part that cares about vote extensions (in live consensus), and other parts that don't (e.g. block sync). In #8375, a boolean flag was introduced into the `Vote.ValidateBasic` method to tell the validation method whether or not to care about these extensions. @tychoish and I spoke about this on Friday and we agreed that it's better to rather have two separate methods for this than introducing a boolean flag.

Again, this PR will form the basis of further work in the direction of resolving #8272 and implementing RFC 017.

Reply:
To further clarify, only the caller will know whether vote extensions matter in their or not.

Reply:
> To further clarify, only the caller will know whether vote extensions matter in their context or not.

Ah, right, I see. ✅

Reply:
I think that requiring vote extensions in the vote message is the right call because they should always be required within consensus. However, we will also need to conditionally require that extensions are present on precommits that were recreated during the switch to consensus. Does this code allow for that?

Reply:
Not specifically the validation logic within the `VoteMessage.ValidateBasic` call, but it is more broadly facilitated by having two distinct validation methods for `Vote`.

Right now, the `CommitToVoteSet` function (which, as I understand it, will be removed in upcoming PRs) uses the `VoteSet.AddVote` function to add votes when reconstructing them in the switch to consensus, which only calls `VerifyWithExtension`: https://github.com/tendermint/tendermint/blob/thane/8272-vote-extension-cleanup/types/vote_set.go#L197

At present this ignores vote extension signatures when we have empty extensions. This will be addressed in subsequent PRs making use of code from #8375.

I've now updated the code to temporarily, while `CommitToVoteSet` still exists, perform basic validation on votes as they're constructed from commits.

Reply:
> Right now, the CommitToVoteSet function (which, as I understand it, will be removed in upcoming PRs)

Perhaps I'm forgetting this discussion, what is the aim in removing that method?

Reply:
I suppose another approach we could consider (probably a separate change if we chose to do so) would be to introduce a separate type for "reconstructed" votes, with the same underlying structure but different validation rules. 

That would let us enforce that "real" votes always require extensions, while "reconstructed" votes would not be required to have them. 

Not that we have to solve this problem with types, but this could help balance those concerns. YMMV.

Reply:
>>    Right now, the CommitToVoteSet function (which, as I understand it, will be removed in upcoming PRs)
>
>Perhaps I'm forgetting this discussion, what is the aim in removing that method?

Right now it's only used in one place in the production code and one test. In the production code it's used when we reconstruct the last commit as we switch from block sync to consensus, so, if I understand correctly, we won't need to reconstruct the last commit if we've saved it along with its vote extensions - we just load it from the relevant store.

> I suppose another approach we could consider (probably a separate change if we chose to do so) would be to introduce a separate type for "reconstructed" votes, with the same underlying structure but different validation rules.

This was my original approach before submitting the current one (technically I had `Vote` and `VoteWithoutExtension`). I also went as far as to define two entirely different structures with mostly the same fields, except `VoteWithoutExtension` didn't have the `Extension` or `ExtensionSignature` fields. Unfortunately this would have involved substantially more refactoring than the current approach, so I abandoned it.

But I agree in principle.